### PR TITLE
[checkpoint] use gather_tensor in checkpoint and update its unit test

### PR DESCRIPTION
**Pull request by 1SAA**
- only gathers tensors to rank0 to reduce memory usage
- corrected and polished colo_tensor's checkpointing unit test

### CODE REVIEW

File: colossalai/utils/checkpoint/module_checkpoint.py [hunk 1]
```diff
@@ -22,37 +22,52 @@ def save_checkpoint(dire: str,
...
+
+    if rank == 0:
+        # sanity check
+        for k, v in model_state.items():
             if isinstance(v, ColoTensor):
+                assert v.save_ready
                 assert v.is_replicate()
+                delattr(v, 'save_ready')
+        # model saving
+        save_state = {'epoch': epoch, 'model': model_state}
```
Comment (feifeibear) [inline colossalai/utils/checkpoint/module_checkpoint.py:41]: In the next PR. You can merge model and optim in a single file.
```
torch.save({
            'epoch': EPOCH,
            'model_state_dict': net.state_dict(),
            'optimizer_state_dict': optimizer.state_dict(),
            'loss': LOSS,
            }, PATH)
```
like 
https://pytorch.org/tutorials/recipes/recipes/saving_and_loading_a_general_checkpoint.html

File: colossalai/utils/checkpoint/utils.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    """Make colo_tensor replicated when the rank is 0
+    """
+    if not colo_tensor.is_replicate():
+        pg = colo_tensor.get_process_group()
+        # for the group which contains rank 0
+        if pg.tp_rank_list()[0] == 0:
+            old_dist_spec = colo_tensor.dist_spec
+            colo_tensor.to_replicate_()
+            if dist.get_rank() != 0:
+                colo_tensor.set_dist_spec(old_dist_spec)
```
Comment (feifeibear) [inline colossalai/utils/checkpoint/utils.py:17]: This line triggers collective communication. 
Will there be potential blocking if rank 0 is excluded?

File: tests/test_utils/test_colo_checkpoint.py [hunk 3]
```diff
@@ -155,43 +159,33 @@ def _run_checkpoint(model_name, init_spec_func, use_ddp, use_mp_reload, test_sch
...
     if rank == 0:
         remove('./checkpoint')
+    dist.barrier()
 
 
 def run_dist(rank, world_size, port, use_ddp, use_mp_reload, test_scheduler):
     colossalai.launch(config={}, rank=rank, world_size=world_size, host='localhost', port=port, backend='nccl')
     pg = ProcessGroup(tp_degree=world_size)
-    for model_name in ['simple_net', 'bert']:
+    # TODO(haichen) add BERT in the test
```
Comment (feifeibear) [inline tests/test_utils/test_colo_checkpoint.py:186]: Inside a DP group, the input is replicated?

Comment (1SAA) [reply tests/test_utils/test_colo_checkpoint.py:186]: Depends on which model is using, we do not unite the standard now.

**Review by feifeibear — APPROVED** [review]

**Pull request merged**